File: Heap/2542. Maximum Subsequence Score/solution.py
class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        arr = sorted(zip(nums1, nums2), key=lambda x: -x[1])
        
        minHeap = []
        res = currSum = 0
        for num, currMin in arr:
            heapq.heappush(minHeap, num)
            currSum += num
            if len(minHeap) > k:
                currSum -= heapq.heappop(minHeap)
            if len(minHeap) == k:
                res = max(res, currSum * currMin)
        return res

File: Heap/2542. Maximum Subsequence Score/solution.cpp
using pii = pair<int, int>;

class Solution {
public:
    long long maxScore(vector<int>& nums1, vector<int>& nums2, int k) {
        int n = nums1.size();
        if (k > n || n == 0 || k == 0) return 0;

        vector<pair<int,int>> pairs;
        for (int i=0; i<n; i++) {
            pairs.push_back({nums2[i], nums1[i]});
        }

        sort(pairs.begin(), pairs.end(), greater{});

        long long sum=0;
        priority_queue<int, vector<int>, greater<int>> min_heap;
        for (int i=0; i<k; i++) {
            sum += pairs[i].second;
            min_heap.push(pairs[i].second);
        }
        long long res = pairs[k-1].first * sum;

        for (int i=k; i<n; i++) {
            min_heap.push(pairs[i].second);
            sum += pairs[i].second;

            sum -= min_heap.top();
            min_heap.pop();

            res = max(res, pairs[i].first * sum);
        }
        return res;
    }
};
